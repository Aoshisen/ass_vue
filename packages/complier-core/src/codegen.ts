import { isString } from "@ass-vue/shared";
import { NodeTypes } from "./ast";
import {
  CREATE_ELEMENT_VNODE,
  TO_DISPLAY_STRING,
  helperMapName,
} from "./runtimeHelpers";

export function generator(ast) {
  const context = createCodegenContext();
  const { push } = context;
  const args = ["_ctx", "_cache"];

  genFunctionPreamble(ast, context);

  const functionName = "render";
  const signature = args.join(", ");

  push(`function ${functionName}(${signature}) {`);
  push("return ");
  genNode(ast.codegenNode, context);
  push("}");
  return {
    code: context.code,
  };
}

function genFunctionPreamble(ast, context) {
  const { push } = context;
  const VueBinging = "Vue";

  const aliasHelper = (s) => `${helperMapName[s]}:_${helperMapName[s]}`;
  if (ast.helpers.length) {
    push(`const {${ast.helpers.map(aliasHelper).join(", ")}}=${VueBinging}`);
  }
  push("\n");
  push("return ");
}

function genNode(node: any, context: any) {
  // console.log("genNode>>>>>>", node);
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpress(node, context);
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    default:
      break;
  }
}

function genElement(node, context) {
  const { push, helper } = context;
  const { tag, children, props } = node;
  push(`${helper(CREATE_ELEMENT_VNODE)}(`);
  // genNode(children, context);
  const nodeList = genNullable([tag, props, children]);
  genNodeList(nodeList, context);
  push(")");
}

function genNullable(args: any) {
  return args.map((arg) => arg || "null");
}

function genNodeList(nodes, context) {
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isString(node)) {
      context.push(node);
    } else {
      genNode(node, context);
    }
    if (i < nodes.length - 1) {
      context.push(",");
    }
  }
}
function genText(node, context) {
  const { push } = context;
  push(`'${node.content}'`);
}

function genInterpolation(node, context) {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(")");
}

function createCodegenContext() {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
    helper(key) {
      return `_${helperMapName[key]}`;
    },
  };
  return context;
}

function genExpress(node: any, context: any) {
  context.push(`${node.content}`);
}

function genCompoundExpression(node: any, context: any) {
  const { push } = context;
  for (let i = 0; i < node.children.length; i++) {
    const child = node.children[i];
    if (isString(child)) {
      push(child);
    } else {
      genNode(child, context);
    }
  }
}
