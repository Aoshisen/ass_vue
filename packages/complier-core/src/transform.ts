import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";

//传入nodeTransform 属性到options 里面，就会执行transform 然后改造我们的node
export function transform(root: any, options: any = {}) {
  const context = createTransformContext(root, options);
  travelNode(root, context);
  createRootCodegen(root, context);
  root.helpers = [...context.helpers.keys()];
}

//rootCodegen Node for codegen
function createRootCodegen(root: any, context) {
  const { children } = root;
  const child = children[0];
  if (child.type === NodeTypes.ELEMENT && child.codegenNode) {
    const codegenNode = child.codegenNode;
    root.codegenNode = codegenNode;
  } else {
    root.codegenNode = child;
  }
}

function createTransformContext(root: any, options: any) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1);
    },
  };
  return context;
}

function travelNode(node: any, context) {
  // console.log("travelNode>>>>>>>>>>>",node);
  //因为调用的时候会去改变我们的text 类型的结构，变成我们的复合类型，所以我们设计一下，先把我们的 复合类型函数收集起来，然后等我们的text 节点类型处理完成之后然后再去处理复合类型的函数

  const exitFn: any = [];

  const nodeTransforms = context.nodeTransforms;

  for (let i = 0; i < nodeTransforms.length; i++) {
    const nodeTransform = nodeTransforms[i];
    const onExit = nodeTransform(node, context);
    if (onExit) {
      exitFn.push(onExit);
    }
  }
  // console.log("travelNode>>>>end",node)
  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING);
      break;
    case NodeTypes.ELEMENT:
    case NodeTypes.ROOT:
      travelChildren(node, context);
      break;
    default:
      break;
  }

  //这里的设计非常巧妙，刚开始的时候我们的i是length ，然后循环一次i减少1，然后我们如果到0的时候也不会越界；
  let i = exitFn.length;
  while (i--) {
    exitFn[i]();
  }
}

function travelChildren(parent: any, context: any) {
  parent.children.forEach((node) => {
    travelNode(node, context);
  });
}
