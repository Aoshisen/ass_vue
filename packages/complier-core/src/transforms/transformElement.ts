import { NodeTypes, createVNodeCall } from "../ast";

export default function transformElement(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      // context.helper(CREATE_ELEMENT_VNODE);

      //tag
      const vnodeTag = `'${node.tag}'`;

      //props
      let vnodeProps;

      //children
      const children = node.children;

      const vnodeChildren = children[0];

      node.codegenNode = createVNodeCall(
        context,
        vnodeTag,
        vnodeProps,
        vnodeChildren
      );
    };
  }
}
