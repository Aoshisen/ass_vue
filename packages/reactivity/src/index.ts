export {ref,proxyRefs,isRef,unRef} from "./ref"
export {shallowReadonly,isProxy,isReactive,isReadonly,reactive,readonly} from "./reactive"
export {effect} from "./effect"
