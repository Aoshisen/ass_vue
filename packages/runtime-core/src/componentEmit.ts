import { camelize, toHandlerKey } from "@ass-vue/shared";

export function emit(instance, event, ...arg) {
  const { props } = instance;
  //TPP
  // 先去写一个特定的行为然后再重构成一个通用的行为

  const handlerName = toHandlerKey(camelize(event));
  const handler = props[handlerName];
  handler && handler(...arg);
}
