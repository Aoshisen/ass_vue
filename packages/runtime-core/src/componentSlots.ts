import { shapeFlags } from "@ass-vue/shared";

export function initSlots(instance, childrenObject) {
  //     instance.slots = Array.isArray(children) ? children : [children];
  const { vnode } = instance;
  if (vnode.shapeFlag & shapeFlags.SLOT_CHILDREN) {
    normalizeSlotObject(instance.slots, childrenObject);
  }
}

function normalizeSlotObject(slots, childrenObject) {
  for (const key in childrenObject) {
    const slot = childrenObject[key];
    slots[key] = (props) => normalizeSlotValue(slot(props));
  }
}

function normalizeSlotValue(slot) {
  return Array.isArray(slot) ? slot : [slot];
}
