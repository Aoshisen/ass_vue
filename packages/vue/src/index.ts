// ass-vue 的出口文件
export * from "@ass-vue/runtime-dom"
import { baseCompile } from "@ass-vue/complier-core";
import * as runtimeDom from "@ass-vue/runtime-dom";
import { registerRuntimeCompiler } from "@ass-vue/runtime-dom";
function compileToFunction(template) {
  const { code } = baseCompile(template);
  //这里的code 就是我们代码compile 生成的代码；第一个参数是我们的函数参数
  const render = new Function("Vue", code)(runtimeDom);
  return render;
}

registerRuntimeCompiler(compileToFunction);
