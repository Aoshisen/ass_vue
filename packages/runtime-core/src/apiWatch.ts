import { ReactiveEffect } from "../../reactivity/src/effect";

import { queueFlushCb } from "./scheduler";
export function watchEffect(source) {
  // source 函数会被添加到组件渲染前调用
  let cleanup;
  function job(cleanupFn) {
    effect.run();
  }

  const onCleanup = function (cleanupFn) {
    cleanup = effect.onStop=() => { cleanupFn() }
  };

  function getter() {
    if (cleanup) {
      cleanup();
    }
    source(onCleanup);
  }
  const effect = new ReactiveEffect(getter, () => {
    queueFlushCb(job);
    // effect.run();
  });
  //一上来的时候这个source 是要被调用的
  effect.run();
  return () => {
    effect.stop();
  };
}
