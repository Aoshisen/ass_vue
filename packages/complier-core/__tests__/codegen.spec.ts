import { generator } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import transformElement from "../src/transforms/transformElement";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformText } from "../src/transforms/transformText";

// import { describe, it, expect } from "vitest";
describe("codeGen", () => {
  it("string", () => {
    const ast = baseParse("hi");

    transform(ast);

    const { code } = generator(ast);

    expect(code).toMatchSnapshot();
  });

  it("interpolation", () => {
    const ast = baseParse("{{message}}");

    transform(ast, { nodeTransforms: [transformExpression] });

    const { code } = generator(ast);

    expect(code).toMatchSnapshot();
  });

  it("element", () => {
    const ast: any = baseParse("<div>hi,{{message}}</div>");
    transform(ast, {
      nodeTransforms: [transformExpression, transformElement, transformText],
    });
    console.log("ast", ast.codegenNode.children);
    const { code } = generator(ast);
    expect(code).toMatchSnapshot();
  });
});
