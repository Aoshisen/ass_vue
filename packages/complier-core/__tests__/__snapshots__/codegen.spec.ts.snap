// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codeGen > element 1`] = `
"const {toDisplayString:_toDisplayString, createElementVNode:_createElementVNode}=Vue
return function render(_ctx, _cache) {return _createElementVNode('div',null,'hi,' + _toDisplayString(_ctx.message))}"
`;

exports[`codeGen > interpolation 1`] = `
"const {toDisplayString:_toDisplayString}=Vue
return function render(_ctx, _cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`codeGen > string 1`] = `
"
return function render(_ctx, _cache) {return 'hi'}"
`;

exports[`codeGen element 1`] = `
"const {toDisplayString:_toDisplayString, createElementVnode:_createElementVnode}=Vue
return function render(_ctx, _cache) {return _createElementVnode('div',null,'hi,' + _toDisplayString(_ctx.message))}"
`;

exports[`codeGen interpolation 1`] = `
"const {toDisplayString:_toDisplayString}=Vue
return function render(_ctx, _cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`codeGen string 1`] = `
"
return function render(_ctx, _cache) {return 'hi'}"
`;
